File: android/app/src/main/java/com/safetylink/app/receivers/WatchdogAlarmReceiver.java
package com.safetylink.app.receivers;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.util.Log;
import com.safetylink.app.services.WatchdogService;

public class WatchdogAlarmReceiver extends BroadcastReceiver {
    private static final String TAG = "SafetyLink::WatchdogAlarm";

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.d(TAG, "Watchdog Ping Received. Validating service state.");
        Intent serviceIntent = new Intent(context, WatchdogService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            try {
                context.startForegroundService(serviceIntent);
            } catch (Exception e) {
                Log.e(TAG, "Failed to start FGS", e);
            }
        } else {
            context.startService(serviceIntent);
        }
    }
}

File: app/src/main/java/com/example/WatchdogAlarmReceiver.kt
package com.example

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.util.Log

/**
 * WatchdogAlarmReceiver
 *
 * Receives the periodic AlarmManager broadcasts and guarantees that WatchdogService is running.
 * Reference: Manifest.permission.SCHEDULE_EXACT_ALARM from https://developer.android.com/reference/android/Manifest.permission
 */
class WatchdogAlarmReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent?) {
        val action = intent?.action ?: return
        Log.i("WatchdogAlarmReceiver", "Received watchdog backup alarm broadcast action: $action")

        if (action == WatchdogService.ACTION_WATCHDOG_CHECK) {
            try {
                val serviceIntent = Intent(context, WatchdogService::class.java)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    context.startForegroundService(serviceIntent)
                } else {
                    context.startService(serviceIntent)
                }
                Log.i("WatchdogAlarmReceiver", "WatchdogService started or validated via Alarm trigger.")
            } catch (e: Exception) {
                Log.e("WatchdogAlarmReceiver", "Failed to launch WatchdogService from alarm tick: ${e.message}", e)
            }
        }
    }
}
